### DISCUSSION TITLE
ObjectId.parser() instand ->  UUID.parser()

### DISCUSSION THREAD

**Discussion by addame2** [pointfreeco/vapor-routing#13] [Q&A]
ObjectId is from mongo db
```swift
public let wordsRouter = OneOf {
    Route(.case(WordsRoute.words)) {
        Path { "words2" }
    }
    
    Route(.case(WordsRoute.word)) {
        Path { "words2" }
        Path { Parse(.string) } // this how do it parse this like UUID.parser() *
        wordRouter
    }
}

func wordHandher(
    request: Request,
    wordID: String,
    route: WordRoute
) async throws -> AsyncResponseEncodable {
    switch route {
    case .fetch:
        guard let id = ObjectId.init(wordID) // * so I can remove this line 
        else {
            struct BadObjectId: Error {}
            throw BadObjectId()
        }
        
        return WordResponse(_id: id, englishWord: "english Word", englishDefinition: "English Definition", isReadFromNotification: false, isReadFromView: false, level: .beginner, url: router.url(for: .words(.word(id.hexString, .fetch))))
}

```

**Comment by stephencelis** [comment]
Hi @addame2, because this is not a bug and more of a feature request or question about usage, I'm going to convert it to a discussion.

Can you provide examples or links to the exact format you want to parse, and the exact data type you want to parse to? For example, what MongoDB library are you using? With that info, folks can try to help you cook up an appropriate parser.

**Reply by addame2** [reply]
thanks for quick reply @stephencelis  I am using this package https://github.com/vapor/fluent-mongo-driver.git inside this package use this package BSON
https://github.com/orlandos-nl/BSON/blob/40956d97c36aa166b8baea5add897864e1568a78/Sources/BSON/Types/ObjectId.swift#L9
its look like UUID

**Reply by stephencelis** [reply]
@addame2 Can you show an example input that you would like to parse into an `ObjectId`?

**Reply by addame2** [reply]
yes @stephencelis  http://0.0.0.0:8080/api2/words2/6235796979bf0df21eba4ed7 if you mean url
here I am use now string here is my handler 

```
public let wordsRouter = OneOf {
    Route(.case(WordsRoute.words)) {
        Path { "words2" }
    }
    
    Route(.case(WordsRoute.word)) {
        Path { "words2" }
        Path { Parse(.string) } 
       // this how do it parse this like UUID.parser() *
       // but ideal way it ObjectId.parser() YES?
        wordRouter
    }
}
```

**Reply by stephencelis** [reply]
@addame2 I think you could define the parser as follows:

```swift
let objectIdParser = ParsePrint {
  Prefix { $0.isHexDigit }.map(
    .convert(
      apply: { ObjectId($0) }
      unapply: { $0.hexString }
    }
  )
}

public let wordsRouter = OneOf {
  Route(.case(WordsRoute.words)) {
    Path { "words2" }
  }
    
  Route(.case(WordsRoute.word)) {
    Path {
      "words2"
       Parse(.string)
       objectIdParser
    }
    wordRouter
  }
}

```

**Reply by addame2** [reply]
<img width="1006" alt="Screenshot 2022-06-22 at 22 59 55" src="https://user-images.githubusercontent.com/85946050/175125526-223a1830-5b3d-4ef3-83d3-652442b57c60.png">

**Reply by stephencelis** [reply]
Ah, we need to further convert the `Substring` parsed to a `String`. Try:

```swift
let objectIdParser = ParsePrint {
  Prefix<Substring> { $0.isHexDigit }
    .map(.string)
    .map(
      .convert(
        apply: { ObjectId($0) },
        unapply: { $0.hexString }
      )
    )
}
```

I'm not sure if the `<Substring>` is needed, and we could maybe offer a `.convert` method to avoid the double-`map`, but hopefully this unblocks you!

**Comment by addame2** [comment]
<img width="1020" alt="Screenshot 2022-06-24 at 19 04 54" src="https://user-images.githubusercontent.com/85946050/175574280-a90db8f0-a672-41d4-bf47-a63710229b1e.png">

**Reply by saroar** [reply]
@stephencelis can you help here?

**Comment by nashysolutions** [comment]
I'm pursuing a similar goal.

```swift

Get: room named 'zeus'
/rooms/zeus

struct RoomNameParser<Input: StringProtocol>: Parser where Input.SubSequence == Input {
    
    func parse(_ path: inout Input) throws -> String {
        guard let name = path.removingPercentEncoding else {
            throw RoomNameParsingError.percentEncoding
        }
        guard name.rangeOfCharacter(from: .alphanumerics.inverted) == nil else {
            throw RoomNameParsingError.invalidChars
        }
        guard RoomNameValidator(value: name).isValid else {
            throw RoomNameParsingError.invalidCharLength
        }
        return name.lowercased()
    }
}

extension RoomNameParser: ParserPrinter where Input: PrependableCollection {
    
    func print(_ output: String, into input: inout Input) throws {
        input.prepend(contentsOf: output)
    }
}

extension String {
    
    static func nameParser<Input>() -> RoomNameParser<Input> {
        .init()
    }
}
```

In my route handler, I have this.

```swift
Route(.case(RoomsRoute.room)) {
    Path { String.nameParser() }
    roomRoute
}
```

All building fine, but.

```
Routing error: multiple failures occurred

error: unexpected input
 --> input:1:8
1 | /rooms/zeus
  |        ^ expected end of input
  |        ^ expected end of input
```

Previously I had a UUID value instead of `zeus` and it worked fine when I used the following

```swift
Route(.case(RoomsRoute.room)) {
    Path { UUID.parser() }
    roomRoute
}
```

@stephencelis you have any guidance on why I'm getting `unexpected end of input` please?

**Reply by saroar** [reply]
why you dont use just string?

**Reply by nashysolutions** [reply]
Could you elaborate a bit more please?

**Reply by saroar** [reply]
This way you can use string 
https://github.com/AddaMeSPB/NWSharedModels/blob/main/Sources/NWSharedModels/Route/UsersRoute.swift#L10

**Reply by nashysolutions** [reply]
Lol ye that will do haha. Thank you.
